spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepository.java

```diff
@@ -0,0 +1,217 @@
+				? serverProperties.getDefaultProfile() : specifiedProfile;
+		final String label = StringUtils.isEmpty(specifiedLabel)
+				? serverProperties.getDefaultLabel() : specifiedLabel;
+		StringBuilder objectKeyPrefix = new StringBuilder(application).append("-")
+				.append(profile);
+		if (!StringUtils.isEmpty(label)) {
+			objectKeyPrefix.append("-").append(label);
+		}
+		final Environment environment = new Environment(application, profile);
+		environment.setLabel(label);
+		final S3ObjectIdBuilder s3ObjectIdBuilder = new S3ObjectIdBuilder()
+				.withBucket(bucketName);
+		S3ConfigFile s3ConfigFile = getS3ConfigFile(s3ObjectIdBuilder,
+				objectKeyPrefix.toString());
+		if (s3ConfigFile == null) {
+			throw new NoSuchRepositoryException(
+					"No such repository: (" + s3ObjectIdBuilder
+							.withKey(objectKeyPrefix.toString() + "(.properties | .yml)")
+							.build().toString() + ")");
+		}
+		environment.setVersion(s3ConfigFile.getVersion());
+		final Map config = s3ConfigFile.read();
+		config.putAll(serverProperties.getOverrides());
+		environment.add(new PropertySource(application, config));
+		return environment;
+	}
+
+	private S3ConfigFile getS3ConfigFile(S3ObjectIdBuilder s3ObjectIdBuilder,
+			String keyPrefix) {
+		try {
+			final S3Object properties = s3Client.getObject(new GetObjectRequest(
+					s3ObjectIdBuilder.withKey(keyPrefix + ".properties").build()));
+			return new PropertyS3ConfigFile(properties.getObjectMetadata().getVersionId(),
+					properties.getObjectContent());
+		}
+		catch (Exception eProperties) {
+			try {
+				final S3Object yaml = s3Client.getObject(new GetObjectRequest(
+						s3ObjectIdBuilder.withKey(keyPrefix + ".yml").build()));
+				return new YamlS3ConfigFile(yaml.getObjectMetadata().getVersionId(),
+						yaml.getObjectContent());
+			}
+			catch (Exception eYaml) {
+				return null;
+			}
+		}
+	}
+
+}
+
+abstract class S3ConfigFile {
+
+	private final String version;
+
+	protected S3ConfigFile(String version) {
+		this.version = version;
+	}
+
+	String getVersion() {
+		return version;
+	}
+
+	abstract Map<?, ?> read();
+
+}
+
+class PropertyS3ConfigFile extends S3ConfigFile {
+
+	final InputStream inputStream;
+
+	PropertyS3ConfigFile(String version, InputStream inputStream) {
+		super(version);
+		this.inputStream = inputStream;
+	}
+
+	@Override
+	public Map<?, ?> read() {
+		Properties props = new Properties();
+		try (InputStream in = inputStream) {
+			props.load(in);
+		}
+		catch (IOException e) {
+			throw new IllegalStateException("Cannot load environment", e);
+		}
+		return props;
+	}
+
+}
+
+class YamlS3ConfigFile extends S3ConfigFile {
+
+	final InputStream inputStream;
+
+	YamlS3ConfigFile(String version, InputStream inputStream) {
+		super(version);
+		this.inputStream = inputStream;
+	}
+
+	@Override
+	public Map<?, ?> read() {
+		final Yaml yaml = new Yaml();
+		final Map<String, String> properties = new HashMap<>();
+		try (InputStream in = inputStream) {
```

Comment:
I think we need to close this `InputStream` at some point

Reply:
This is [try with resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) and will be closed automatically.

Reply:
try w/ resources will close it automatically.